### PR TITLE
Fixes for services playback per review 1

### PR DESCRIPTION
**Pull request by MichaelOrlov**
- Relates https://github.com/ros2/rosbag2/pull/1481
## Made the following changes per review:
- Cleanup and optimization in mcap_storage.cpp
  - Rewrite topic_filter to make clean and concise implementation.
  - Also addressed multiple performance related issues in topic_filter.
  - Delete `ROSBAG2_STORAGE_MCAP_HAS_STORAGE_FILTER_TOPIC_REGEX` compilation flag since it is become absolute on Rolling, and we are not using anymore the same version of the mcap storage plugin for older ROS 2 distros.
- Move mcap topic_filter tests to a separate file
- Add `TestResetFilter`, `CanSelectWithTopicsListOnly` and `CanSelectWithServiceEventsListOnly` tests to the mcap storage
- Add `TestResetFilter`, `CanSelectWithTopicsListOnly` and `CanSelectWithServiceEventsListOnly` tests to the sqlite3 storage
- Cleanups and renames in the sqlite_storage.cpp
- Move SQLite3 topic_filter tests to a separate file
- Bugfix in sqlite topic filters when service_events or topics lists empty
- Changed API for `async_send_request(..)`, `is_include_request_message(..)` and `get_msg_event_type(..)` to use `rcl_serialized_message_t` instead of the `rclcpp::SerializedMessage` to avoid 2 extra message copy.
- Misc minor fixes from previous rounds of review and new findings

### CODE REVIEW

**Comment by MichaelOrlov** [comment]
@Barry-Xu-2018 @fujitatomoya Please review my changes per review.  
That is not all, but at least the first part.

**Comment by MichaelOrlov** [comment]
@Barry-Xu-2018 Please note that newly added tests in the `rosbag_transport` for services replay are tend to be flaky and fails on the `Test rosbag2 / build_and_test (pull_request)` because:
1. Suboptimal implementation in the 
   ```
     void run_services()
      {
        auto spin = [this]() {
            rclcpp::executors::SingleThreadedExecutor exec;
            exec.add_node(pub_node_);
  
            while (!exit_) {
              exec.spin_some();
              std::this_thread::sleep_for(std::chrono::milliseconds(50));
            }
          };
  
        thread_ = std::thread(spin);
      }
   ```
     IMO we shall not do  
   ```cpp
           exec.spin_some();
           std::this_thread::sleep_for(std::chrono::milliseconds(50));
    ```
     because we can miss services during sleep. Need to wait on condition variable and use `exec.cancel()` API please refer to the `rosbag2_py::_transport.cpp`  or https://github.com/ros2/rosbag2/blob/rolling/rosbag2_transport/test/rosbag2_transport/test_play_services.cpp for example.
2. In each test itself there is a race condition between thread that starts spinning receiving services node in test and between ` player->play();` operation. It turns out that on CI  `player->play();` can start playing early than the receiver side will be ready to receive. To mitigate this issue need to start playback in paused mode via `play_options_.start_paused = true` then wait for services to be discoverable on the receiver side via `wait_for_service(timeout)` API. Please see how it is done in the https://github.com/ros2/rosbag2/blob/rolling/rosbag2_transport/test/rosbag2_transport/test_play_services.cpp for example.

cc: @@fujitatomoya

File: rosbag2_storage_mcap/src/mcap_storage.cpp
Comment (Barry-Xu-2018) [inline rosbag2_storage_mcap/src/mcap_storage.cpp]: There is duplicated code for regex.   
Can it be simplified like this? 

```suggestion
  // Both white list and regex are empty. i.e. equivalent to include all
  if (white_list.empty() && regex.empty()) {
    return true;
  }

  if (std::find(white_list.begin(), white_list.end(), topic_name) != white_list.end()) {
      return true;
  }
  
  if (!regex.empty()) {
    std::smatch m;
    std::string topic_string(topic_name);
    std::regex re(regex);
    if (std::regex_match(topic_string, m, re)) {
      return true;
    }
  }

  return false;
```

File: rosbag2_storage_mcap/src/mcap_storage.cpp
Comment (Barry-Xu-2018) [inline rosbag2_storage_mcap/src/mcap_storage.cpp]: I think is_topic_in_black_list_or_exclude_regex() should be checked at the first.    
Users might use regex to include topics, but want to remove some using an exclude_list. If we first execute is_topic_selected_by_white_list_or_regex(), we still need to execute is_topic_in_black_list_or_exclude_regex().

```suggestion
    // if topic not found in exclude list or regex_to_exclude
    if (!is_topic_in_black_list_or_exclude_regex(topic, exclude_list,
                                                 storage_filter_.regex_to_exclude))
    {
      // if topic found in include list or regex
      if (is_topic_selected_by_white_list_or_regex(topic, include_list, storage_filter_.regex)) {
        return true;
      }
    }

    return false;
```

File: rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp
Comment (Barry-Xu-2018) [inline rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp]: storage_filter.services_events.empty() does not mean that the user has not set conditions for services. Users can set it via regex (Regex affects the selection of both topic and service event topic simultaneously.) .  So I think we cannot add "OR  regex_for_all_service_events_str" here.

File: rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp
Comment (Barry-Xu-2018) [inline rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:601]: Like above comments. storage_filter.topics.empty() return true, which doesn't mean users not set condition for topic. He can set it via regex.

File: rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp
Comment (Barry-Xu-2018) [inline rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:607]: Regex maybe set. So we cannot think it is an exception while storage_filter.services_events.empty() == true and storage_filter.topics.empty() == true.

**Review by Barry-Xu-2018 — COMMENTED** [review]
@MichaelOrlov  
Thank you for your modifications. I have no comments to most of them.   
Regarding the filtering process, I would like to discuss with you.

**Comment by Barry-Xu-2018** [comment]
After this PR is merged, I will update codes base on your review comments https://github.com/Barry-Xu-2018/rosbag2/pull/5#issuecomment-2019150578

**Review by fujitatomoya — COMMENTED** [review]
@MichaelOrlov thanks for the changes.

i think test replacements is much cleaner for maintenance. I am good to go with @Barry-Xu-2018 's comments answered.

File: rosbag2_storage_mcap/src/mcap_storage.cpp
Comment (MichaelOrlov) [reply rosbag2_storage_mcap/src/mcap_storage.cpp]: Good catch. Make sense to me.

File: rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp
Comment (MichaelOrlov) [reply rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:601]: In this particular case is safe to include `all_topics` because we will get to this line only if `regex_filter_str` is empty.
 if (!topics_filter_str.empty() && !regex_filter_str.empty()) {
    where_conditions.push_back("(" + topics_filter_str + " OR " + regex_filter_str + ")");
  } else if (!topics_filter_str.empty()) {  // Note: regex_filter_str is empty at the same time
//   ...
  }
I will add a comment with a note.

File: rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp
Comment (MichaelOrlov) [reply rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:607]: The same situation as above. 
In this particular case, we will get to this line only if regex_filter_str is empty.

File: rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp
Comment (MichaelOrlov) [reply rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp]: Good catch! I will fix and add test coverage for these use cases.

File: rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp
Comment (Barry-Xu-2018) [reply rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:601]: Thanks. I was wrong.

File: rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp
Comment (Barry-Xu-2018) [reply rosbag2_storage_sqlite3/src/rosbag2_storage_sqlite3/sqlite_storage.cpp:607]: Yes.

**Comment by MichaelOrlov** [comment]
@Barry-Xu-2018 I've addressed your comments. Please review/verify.

**Comment by MichaelOrlov** [comment]
@Barry-Xu-2018 If you have no concerns or comments in regard to this PR please merge it on your branch or cherrypick commits from my branch.

**Review by fujitatomoya — APPROVED** [review]
@MichaelOrlov thanks for the effort. lgtm! i will leave this to @Barry-Xu-2018

**Comment by Barry-Xu-2018** [comment]
LGTM  
I merge them.   
Thank you so much. @MichaelOrlov

**Pull request merged**
